log/loggers.go

```diff
@@ -0,0 +1,193 @@
+)
+
+// LoggerInfo is a structure that should be used when creating a logger.
+// It contains parameters about how to log (with time, colors, ...) and
+// embeds the Logger interface, which should define how the logger should log.
+type LoggerInfo struct {
+	// These are information-debugging levels that can be turned on or off.
+	// Every logging greater than 'debugLvl' will be discarded . So you can
+	// Log at different levels and easily turn on or off the amount of logging
+	// generated by adjusting the 'debugLvl' variable.
+	debugLvl int
+	// If 'showTime' is true, it will print the time for each line displayed
+	// by the logger.
+	showTime bool
+	// If 'useColors' is true, logs will be colored (defaults to monochrome
+	// output). It also controls padding, since colorful output is higly
+	// correlated with humans who like their log lines padded.
+	useColors bool
+	// If 'padding' is true, it will nicely pad the line that is written.
+	padding bool
+	// When using LoggerInfo, we need to be able to act as a Logger.
+	Logger
+}
+
+// Logger is the interface that specifies how loggers
+// will receive and display messages.
+type Logger interface {
+	Log(level int, msg string, l *LoggerInfo)
+	Close()
+}
+
+const (
+	// DefaultStdDebugLvl is the default debug level for the standard logger
+	DefaultStdDebugLvl = 1
+	// DefaultStdShowTime is the default value for 'showTime' for the standard logger
+	DefaultStdShowTime = false
+	// DefaultStdUseColors is the default value for 'useColors' for the standard logger
+	DefaultStdUseColors = false
+	// DefaultStdPadding is the default value for 'padding' for the standard logger
+	DefaultStdPadding = true
+)
+
+var (
+	// concurrent access is protected by debugMut
+	loggers        = make(map[int]*LoggerInfo)
+	loggersCounter int
+)
+
+// RegisterLogger will register a callback that will receive a copy of every
+// message, fully formatted. It returns the key assigned to the logger (used
+// to unregister the logger).
+func RegisterLogger(l *LoggerInfo) int {
+	debugMut.Lock()
+	defer debugMut.Unlock()
+	key := loggersCounter
+	loggers[key] = l
+	loggersCounter++
+	return key
+}
+
+// UnregisterLogger takes the key it was assigned and returned by
+// 'RegisterLogger', closes the corresponding Logger and removes it from the
+// loggers.
+func UnregisterLogger(key int) {
+	debugMut.Lock()
+	defer debugMut.Unlock()
+	if l, ok := loggers[key]; ok {
+		l.Close()
+		delete(loggers, key)
+	}
+}
+
+type fileLogger struct {
+	file *os.File
+}
+
+func (fl *fileLogger) Log(level int, msg string, l *LoggerInfo) {
+	if _, err := fl.file.WriteString(msg); err != nil {
+		panic(err)
+	}
+}
+
+func (fl *fileLogger) Close() {
+	fl.file.Close()
+}
+
+// NewFileLogger creates and registers a logger that writes into the file with
+// the given path and is using the given LoggerInfo (without the Logger).
+// It returns the key assigned to the logger.
+func NewFileLogger(path string, lInfo *LoggerInfo) (int, error) {
```

Comment:
`LoggerInfo` is too prevalent here - it's once in `New.*` and in `Log`. Also, calling `RegisterLogger` in `New*` does more than one would expect from a `New.*` method in go.

Trying to clean up (hopefully the last time, sorry):
`LoggerInfo` does _not_ contain `Logger`. But `fileLogger`, `stdLogger` and `syslogLogger` contain `*LoggerInfo`.
All `New*` take a `*LoggerInfo` as first argument and any specific arguments necessary, they return a `Logger` interface and an error. We usually take the common argument first, and then the specific arguments. This makes it easier to remember the method signature.
`Logger.Log` does only take `lvl, msg` and uses the stored `LoggerInfo`.
`RegisterLogger` takes a `Logger` and returns an `int` for `UnregisterLogger`.

Reply:
Alright! So now, in `lvl`, I will iterate over a list of `Logger`, and not `*LoggerInfo` anymore. This means that I don't have access to things like `debugLvl`, `showTime`, etc... which is annoying since we wanted to do the generic logic of the loggers in there (e.g. checking the debug level), and the specific or more I/O related things in their `Log` method. Or am I missing something and we can easily access the `*LoggerInfo` from the `Logger`?
One solution I see is to make `Logger` implement a `GetLoggerInfo()` method returning a `*LoggerInfo`, and make all loggers implement it and just returning the `*LoggerInfo` they have.
Or maybe there is another way in Go that I don't know?